File: index_model.cpp
#include <algorithm>
#include <iostream>
#include "dataset.h"
#include "util.h"
#include <string>
#include <thread>
#include <list>




using namespace std;


set<string> ban_set = {"1080p", "720p", "3d"};
set<string> prefix_ban_set = {"under", "with", "camera", "full", "black", "pink", "for", "body",
                                        "canon", "ef", "top", "w", "led", "tv", "kit", "lens", "edition",
                                        "only", "pro", "ii", "iii"};
map<string, list<string>> special_model = {{"Canon", {"xt", "xti", "xs", "xsi", "eos m", "eos m2", "1ds"}},
                                                     {"Nikon", {"df", "d1x", "d2x", "d3x", "coolpix A"}},
                                                     {"GoPro", {"Hero 3", "Hero 2", "Hero3+"}},
                                                     {"SVP", {"20MP", "18MP"}},{"Olympus",{"EM 5","EM-5"}},
                                                     {"Hasselblad", {"Hasselblad"}}};



bool word_rule(string word) {
    if (word.size() < 2) {
        return false;
    }
    if (word.size() <= 3 && (word[word.length()-1] == 'x' || word[word.length()-1] == 'X')) {
        return false;
    }
    bool flag = true;
    for (int i = 0; i < word.size()-1; i++) {
        if (!is_digit(word[i])) {
            flag = false;
            break;
        }
    }
    if (flag && is_alphabet(word[word.size() - 1])) {
        return true;
    }
    if (!is_alphabet(word[0])) {
        return false;
    }
    int score = 0;
    for (char c: word) {
        if (is_alphabet(c)) score |= 1;
        else if (is_digit(c)) score |= 2;
        else if (c != '-') score |= 4;
    }
    return score == 3;
}

bool is_valid_prefix(const string& word) {
    if (word.length() < 1) {
        return false;
    }
    for (char c: word) {
        if (!is_alphabet(c)) {
            return false;
        }
    }
    return true;
}

bool is_valid_postfix(string word) {
    if (word.size() < 2 || word.size() > 7) {
        return false;
    }
    for (int i = 0; i < word.size()-1; i++) {
        if (!is_digit(word[i])) {
            return false;
        }
    }
    if (!is_alphabet(word[word.size()-1]) && !is_digit(word[word.size()-1])) {
        return false;
    }
    return str_lower(word)[word.size() - 1] != 'x';
}


void collecting_models(const string& page_title, set<string>& model_set) {
    vector<string> tmp;
    split(page_title, tmp);
    //boost::split(tmp, page_title, boost::is_any_of(" "), boost::token_compress_on);
    //tmp = split(std::move(page_title));

    for (int i = 0; i < tmp.size()-1; i++) {
        for (int step = 0; step <= 1; step ++) {
            string word = tmp[i+step];
            if (word_rule(word) && ban_set.find(str_lower(word)) == ban_set.end()) {
                if (model_set.find(str_upper(word)) != model_set.end()) {
                    return;
                }
                model_set.insert(str_upper(word));
            }
        }
        if (is_valid_prefix(tmp[i]) && (prefix_ban_set.find(str_lower(tmp[i])) == prefix_ban_set.end())
            && is_valid_postfix(tmp[i+1])) {
            string word = tmp[i] + ' ' + tmp[i+1];
            if (model_set.find(str_upper(word)) != model_set.end()) {
                return;
            }
            model_set.insert(str_upper(word));
            return;
        }
    }
}


void matching(map<string, set<string>>& model_index_each, const set<string>& model_set,
              const set<string>& product_set) {
    for (const string& model: model_set) {
        //model_index_each[model] = new unordered_set<string>;
        for (const string& product: product_set) {
            string page_title = all_data[product]->page_title;
            //vector<string> tmp = split(page_title);
            vector<string> tmp;
            //boost::split(tmp, page_title, boost::is_any_of(" "), boost::token_compress_on);
            split(page_title, tmp);
            for (const string& word: tmp) {
                if (str_lower(word) == str_lower(model)) {
                    model_index_each[model].insert(product);
                }
            }
            for (int i = 0; i + 1 < tmp.size(); i++) {
                string word = tmp[i] + " " + tmp[i+1];
                if (str_lower(word) == str_lower(model)) {
                    model_index_each[model].insert(product);
                }
            }
        }
    }
}



inline void merge(string& to_model, string& from_model,
                  map<string, set<string>>& model_index_each) {
    model_index_each[to_model].insert(model_index_each[from_model].begin(), model_index_each[from_model].end());
}



void merge_pairs(const set<string> &model_set,
                 map<string, set<string>> &model_index_each) {
    set<string> remove_models;

    vector<string> models(model_set.begin(), model_set.end());
    sort(models.begin(), models.end());
    for (int i = 0; i < models.size(); i++) {
        for (int j = i+1; j < models.size(); j++) {
            if (str_remove_space(models[i]) == str_remove_space(models[j])) {
                if (remove_models.find(models[i]) == remove_models.end() ||
                    remove_models.find(models[j]) == remove_models.end()) {
                    continue;
                }
                merge(models[i], models[j], model_index_each);
                remove_models.insert(models[j]);
            }
        }
    }
    for (const string& rm_model: remove_models) {
        model_index_each.erase(rm_model);
    }
}


void kill_no_contribute(map<string, set<string>>& model_index_each) {
    list<string> del_list;
    for (auto item: model_index_each) {
        if (item.second.size() < 2) {
            del_list.push_back(item.first);
        }
    }
    for (const string& del_model: del_list) {
        model_index_each.erase(del_model);
    }
}


void index_model_brand(const string& brand, const set<string> product_set) {
    cout << "Indexing model: " << brand << endl;
    map<string, set<string>> each_model_index;
    set<string> model_set;
    // add special
    if (special_model.find(brand) != special_model.end()) {
        for (string model: special_model[brand]) {
            model_set.insert(str_upper(model));
        }
    }

    // find models
    for (const string& product: product_set) {
        string page_title = all_data[product]->page_title;
        collecting_models(page_title, model_set);
    }
    matching(each_model_index, model_set, product_set);
    merge_pairs(model_set, each_model_index);
    kill_no_contribute(each_model_index);
    model_index[brand] = each_model_index;
}

string important_brands[] = {"Canon", "Nikon", "Fujifilm","Sony","Panasonic","Samsung","Olympus"};
void index_model() {
    cout << "Indexing Models!" << endl;
    auto **th = new thread * [4];
    for(int i = 0;i<4;i++){
        th[i] = new thread(index_model_brand,important_brands[i],brand_index[important_brands[i]]);
    }
    for (auto& brand: brand_index) {
        string brand_name = brand.first;
        bool flag = false;
        for(int i=0;i<4;i++){
            if(brand_name == important_brands[i]){
                flag = true;
                break;
            }
        }
        if(flag) continue;
        set<string> set = brand.second;
        index_model_brand(brand_name, set);
    }
    for(int i = 0;i<4;i++){
        th[i]->join();
    }




//    ThreadPool pool(4);
//    std::vector< std::future<int> > results;
//
//
//    for (auto& brand: brand_index) {
//        string brand_name = brand.first;
//        bool flag = false;
//        unordered_set<string> set = brand.second;
//        results.emplace_back( pool.enqueue(index_model_brand, brand_name, set) );
//    }
//
//    for(auto && result: results)
//        std::cout << result.get() << ' ';
//    std::cout << std::endl;



}

File: python version/index_model.py
import dataset

# dataset_path = './brand'
# output_path = './model'

columns_df = ['id', '<page title>']
ban_list = {'1080p': 1, '720p': 1, '3d': 1, '8F': 1}
prefix_ban_list = {'under': 1, 'with': 1, 'camera': 1, 'full': 1, 'black': 1, 'pink': 1, 'for': 1, 'body': 1,
                   'canon': 1, 'ef': 1, 'top': 1, 'w': 1, 'led': 1, 'tv': 1, 'kit': 1, 'lens': 1, 'edition': 1,
                   'only': 1, 'pro': 1, 'ii': 1, 'iii': 1}
special_model = {'Canon': ['xt', 'xti', 'xs', 'xsi', 'eos m', 'eos m2', '1ds'],
                 'Nikon': ['df', 'v1', 'v2', 'd1x', 'd2x', 'd3x', 'coolpix A'],
                 'GoPro': ['Hero 3', 'Hero 2', 'Hero3+'],
                 'Hasselblad': ['Hasselblad']}

vis = {}



def collecting_models(page_title, model_exist, model_list):
    temp = page_title.split(" ")

    for i in range(0, len(temp) - 1):
        word = temp[i]
        if rule1(word) and word.lower() not in ban_list:
            if word.lower() in model_exist:
                return
            model_exist[word.lower()] = 1
            model_list.append(word.upper())
            return

        word = temp[i + 1]
        if rule1(word) and word.lower() not in ban_list:
            if word.lower() in model_exist:
                return
            model_exist[word.lower()] = 1
            model_list.append(word.upper())
            return

        if valid_prefix_test(temp[i]) and temp[i].lower() not in prefix_ban_list and valid_postfix_test(temp[i + 1]):
            word = temp[i] + ' ' + temp[i + 1]
            if word.lower() in model_exist:
                return
            model_exist[word.lower()] = 1
            model_list.append(word.upper())
            return


def valid_prefix_test(word):
    if len(word) < 1:
        return False
    for alphabet in word:
        if not alphabet.isalpha():
            return False
    return True


def valid_postfix_test(word):
    if len(word) < 2 or len(word) > 7:
        return False
    for i in range(0, len(word) - 1):
        if not word[i].isdigit():
            return False
    if not word[len(word) - 1].isalpha() and not word[len(word) - 1].isdigit():
        return False
    if word[len(word) - 1].lower() == 'x':
        return False
    return True


def rule1(word):
    score = 0
    if len(word) < 2:
        return False
    if len(word) <= 3 and word[len(word) - 1].lower() == 'x':
        return False
    Flag = True
    for i in range(0, len(word) - 1):
        if not word[i].isdigit():
            Flag = False
            break
    if word[len(word) - 1].isalpha() and Flag:
        return True
    if not word[0].isalpha():
        return False
    for alphabet in word:
        if alphabet.isalpha():
            score |= 1
        elif alphabet.isdigit():
            score |= 2
        elif alphabet != "-":
            score |= 4
    if score == 3:
        return True

    return False



def matching(brand, model_list):
    products = dataset.brand_index[brand]
    for model in model_list:
        dataset.model_index[brand][model] = []
        for product in products:
            page_title = dataset.all_data[product].get('<page title>')
            temp = page_title.split(' ')
            for word in temp:
                if word.lower() == model.lower():
                    vis[product] = 1
                    dataset.model_index[brand][model].append(product)
            for i in range(0, len(temp) - 1):
                word = temp[i] + ' ' + temp[i + 1]
                if word.lower() == model.lower():
                    vis[product] = 1
                    dataset.model_index[brand][model].append(product)




def kill_non_contribute():
    delete_list = []
    for brand, brand_items in dataset.model_index.items():
        for model, model_items in brand_items.items():
            if len(model_items) < 2:
                delete_list.append((brand, model))
    for delete_brand, delete_model in delete_list:
        del dataset.model_index[delete_brand][delete_model]



def merge(to_model, from_model, brand):
    dataset.model_index[brand][to_model] += dataset.model_index[brand][from_model]
    dataset.model_index[brand][to_model] = list(set(dataset.model_index[brand][to_model]))




def remove_space(model_str):
    model_str = model_str.lower()
    res = ''
    for i in range(0, len(model_str)):
        if model_str[i] == ' ' or model_str[i] == '-':
            continue
        res += model_str[i]
    return res




def merge_pairs(model_list, brand):
    remove_models = []
    for i in range(0, len(model_list)):
        for j in range(i + 1, len(model_list)):
            if remove_space(model_list[i]) == remove_space(model_list[j]):
                merge(model_list[i], model_list[j], brand)
                remove_models.append(model_list[j])
            # remove_path.append('./model/' + brand + '/' + model_list[j] + '.csv')
    for rm_model in set(remove_models):
        del dataset.model_index[brand][rm_model]


def index_model():
    print("Indexing models")
    for brand, product_list in dataset.brand_index.items():
        print("Indexing models: ", brand)
        dataset.model_index[brand] = {}
        model_list = []
        model_exist = {}

        # add special
        if brand in special_model:
            for model in special_model[brand]:
                model_list.append(model.upper())
                model_exist[model.lower()] = 1

        # find_models
        for product in product_list:
            page_title = dataset.all_data[product].get('<page title>')
            collecting_models(page_title, model_exist, model_list)

        # matching
        matching(brand, model_list)

        merge_pairs(model_list, brand)
        kill_non_contribute()




if __name__ == '__main__':
    index_model()
